### PR TITLE
Add --add-host cli option to expose docker container option with the same name.

### PR DESCRIPTION
**Pull request by myukselen**
#### Which issue(s) does this change fix?
None

#### Why is this change necessary?
Docker container support add-host cli option to give additional host to IP mappings. This is useful in doing local testing as we can override DNS resolution and replace some client calls with different endpoints.

#### How does it address the issue?
This change provides --add-host container related option to the SAM cli and it can be used in local invoke, and start api commands.

#### What side effects does this change have?
There is none I am aware with respect to unit-tests.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [x] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by lucashuy** [comment]
Thanks for opening this PR! The changes here seem straightforward to me, but I wanted to check in with the team/security first since involves changing the hostname mapping in the containers.

**Review by lucashuy — CHANGES_REQUESTED** [review]
Sorry about the delay in responses, we have approval to go ahead with this change now.

**Comment by myukselen** [comment]
thanks for the response. I will try to address your requests soon with a revision.

File: samcli/commands/local/cli_common/invoke_context.py [hunk 1]
```diff
@@ -177,6 +180,14 @@ def __init__(
...
         self._container_host = container_host
         self._container_host_interface = container_host_interface
+
+        self._extra_hosts: Optional[Dict] = None
+        if add_host and len(add_host) > 0:
+            self._extra_hosts = {}
+            for host_to_ip in add_host:
+                host, ip = host_to_ip.split(":", maxsplit=1)
+                self._extra_hosts[host] = ip
+
```
[the remark this answers is not in the document]
Comment (myukselen) [reply samcli/commands/local/cli_common/invoke_context.py]: I have added DockerAdditionalHostType for this purpose into `types`. Please suggest any further changes necessary.

File: samcli/commands/local/cli_common/options.py [hunk 2]
```diff
@@ -64,6 +64,12 @@ def local_common_options(f):
             help="IP address of the host network interface that container ports should bind to. "
             "Use 0.0.0.0 to bind to all interfaces.",
         ),
+        click.option(
+            "--add-host",
+            default=None,
+            multiple=True,
+            help="Utilize hostname to IP mapping via docker --add-host flag. Can be multiple."
```
[the remark this answers is not in the document]
Comment (myukselen) [reply samcli/commands/local/cli_common/options.py]: I have added type `DockerAdditionalHostType` and changed the help content.

**Comment by myukselen** [comment]
I made a revision. Please provide any necessary changes to move forward.

**Comment by lucashuy** [comment]
Thanks for making the changes! It looks like Github is saying there is a conflict, can you update your local branch to see what's wrong? When that is done, I can start the checks here on Github to run our tests

File: samcli/local/docker/lambda_container.py [hunk 3]
```diff
@@ -50,8 +50,9 @@ def __init__(
         env_vars=None,
         debug_options=None,
         container_host=None,
-        container_host_interface=None,
+        container_host_interface=DEFAULT_CONTAINER_HOST_INTERFACE,
         extra_hosts=None,
+        container_host_interface=DEFAULT_CONTAINER_HOST_INTERFACE,
```
Comment (lucashuy) [inline samcli/local/docker/lambda_container.py:55]: Looks like this was duplicated accidentally. Additionally, looks like there was another update to a file in the develop branch causing issues with the container file here. I've started the tests anyways (hoping that they all don't get stuck at the duplicate argument). While we wait for the tests and the other conflict to be resolved, I'll raise this to the team to get another reviewer.

Aside from the PR jobs complaining, I don't have any concerns with your logic in the PR!

Comment (myukselen) [reply samcli/local/docker/lambda_container.py]: ok, I deleted the extra line.

File: samcli/commands/_utils/options.py [hunk 4]
```diff
@@ -156,6 +156,21 @@ def remote_invoke_boto_parameter_callback(ctx, param, provided_value):
     return boto_api_parameters
 
 
+def local_add_host_callback(ctx, param, provided_value):
```
Comment (mndeveci) [inline samcli/commands/_utils/options.py:159]: I thought custom type should handle this as well, do you know why we needed a callback function here?

**Comment by lucashuy** [comment]
Hi, it looks like the make command `make format` should be run so that the linter and formatter we have can fix the files in question. This can be done on your local branch and you can run `make pr` to validate that things are fixed correctly too.

**Comment by myukselen** [comment]
Hi, sorry for the delays, I am not familiar with Github PR workflows. I pulled my repo branch and ran `make format` and `make pr` . The test output still shows this one error I have no idea about:

```
FAILED tests/unit/cli/test_import_module_proxy.py::TestImportModuleProxy::test_import_should_succeed_for_a_defined_hidden_package_539_pkg_resources_py2_warn - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```

I just pushed `make format` changes.

File: samcli/commands/_utils/options.py (hunk 4 above)
Comment (myukselen) [reply samcli/commands/_utils/options.py:159]: I have copy/pasted the functionality from the reference `def remote_invoke_parameter_click_option():` and there I saw `RemoteInvokeBotoApiParameterType` is used to parse single argument and `remote_invoke_boto_parameter_callback` callback is utilized to combine them in a dictionary.

If there is a better approach, I can try to refactor it with your guidance.

**Comment by myukselen** [comment]
I just pushed `make schema` changes as I saw this has failed in the workflow.

**Review by lucashuy — APPROVED** [review]
Sorry about the delay in response, thanks for taking the time while we sorted out the comments and the CI checks

**Review by mndeveci — APPROVED** [review]

**Pull request merged**